File: ios/Sources/GhostCrypto/WireFormat.swift
import Foundation

public enum WireFormat {

    public enum WireError: Error {
        case headerTooShort
        case invalidHeaderVersion
        case messageTooShort
        case headerLengthMismatch
    }

    // MARK: - Header: version(1) + dhPubKeyRaw(64) + PN(4, BE) + N(4, BE) = 73 bytes

    public static let headerSize = 73

    public struct ParsedHeader {
        public let version: UInt8
        public let dhPublicKeyRaw: Data  // 64 bytes (x + y, no 04 prefix)
        public let pn: UInt32
        public let n: UInt32
    }

    public static func buildHeader(dhPublicKeyRaw: Data, pn: UInt32, n: UInt32) -> Data {
        var buf = Data(capacity: headerSize)
        buf.append(0x00) // version
        buf.append(dhPublicKeyRaw) // 64 bytes
        var pnBE = pn.bigEndian
        buf.append(Data(bytes: &pnBE, count: 4))
        var nBE = n.bigEndian
        buf.append(Data(bytes: &nBE, count: 4))
        return buf
    }

    public static func parseHeader(_ data: Data) throws -> ParsedHeader {
        guard data.count >= headerSize else { throw WireError.headerTooShort }
        let version = data[data.startIndex]
        guard version == 0x00 else { throw WireError.invalidHeaderVersion }
        let dhPubRaw = data[(data.startIndex + 1)..<(data.startIndex + 65)]
        let pn = data[(data.startIndex + 65)..<(data.startIndex + 69)].withUnsafeBytes {
            $0.loadUnaligned(as: UInt32.self).bigEndian
        }
        let n = data[(data.startIndex + 69)..<(data.startIndex + 73)].withUnsafeBytes {
            $0.loadUnaligned(as: UInt32.self).bigEndian
        }
        return ParsedHeader(version: version, dhPublicKeyRaw: Data(dhPubRaw), pn: pn, n: n)
    }

    // MARK: - Wire message: headerLen(4, BE) + header + nonce(12) + ciphertext + tag(16)

    public struct ParsedMessage {
        public let header: Data
        public let nonce: Data       // 12 bytes
        public let ciphertext: Data
        public let tag: Data         // 16 bytes
    }

    public static func buildMessage(header: Data, nonce: Data, ciphertext: Data, tag: Data) -> Data {
        var buf = Data(capacity: 4 + header.count + nonce.count + ciphertext.count + tag.count)
        var headerLenBE = UInt32(header.count).bigEndian
        buf.append(Data(bytes: &headerLenBE, count: 4))
        buf.append(header)
        buf.append(nonce)
        buf.append(ciphertext)
        buf.append(tag)
        return buf
    }

    public static func parseMessage(_ data: Data) throws -> ParsedMessage {
        guard data.count >= 4 else { throw WireError.messageTooShort }
        let headerLen = data.prefix(4).withUnsafeBytes { $0.loadUnaligned(as: UInt32.self).bigEndian }
        let expected = 4 + Int(headerLen) + 12 + 16
        guard data.count >= expected else { throw WireError.messageTooShort }

        let headerStart = 4
        let headerEnd = headerStart + Int(headerLen)
        let nonceStart = headerEnd
        let nonceEnd = nonceStart + 12
        let tagStart = data.count - 16
        let ctStart = nonceEnd
        let ctEnd = tagStart

        return ParsedMessage(
            header: Data(data[headerStart..<headerEnd]),
            nonce: Data(data[nonceStart..<nonceEnd]),
            ciphertext: Data(data[ctStart..<ctEnd]),
            tag: Data(data[tagStart..<data.count])
        )
    }
}

File: android/crypto/src/main/kotlin/com/kordar/ghostchat/core/crypto/WireFormat.kt
package com.kordar.ghostchat.core.crypto

import java.nio.ByteBuffer
import java.nio.ByteOrder

object WireFormat {

    const val HEADER_SIZE = 73

    data class ParsedHeader(
        val version: Byte,
        val dhPublicKeyRaw: ByteArray,  // 64 bytes
        val pn: Int,
        val n: Int
    )

    /** Build header: version(1) + dhPubKeyRaw(64) + PN(4 BE) + N(4 BE) = 73 bytes */
    fun buildHeader(dhPublicKeyRaw: ByteArray, pn: Int, n: Int): ByteArray {
        val buf = ByteBuffer.allocate(HEADER_SIZE).order(ByteOrder.BIG_ENDIAN)
        buf.put(0x00.toByte()) // version
        buf.put(dhPublicKeyRaw) // 64 bytes
        buf.putInt(pn)
        buf.putInt(n)
        return buf.array()
    }

    fun parseHeader(data: ByteArray): ParsedHeader {
        require(data.size >= HEADER_SIZE) { "Header too short" }
        val buf = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
        val version = buf.get()
        require(version == 0x00.toByte()) { "Invalid header version" }
        val dhPubRaw = ByteArray(64)
        buf.get(dhPubRaw)
        val pn = buf.int
        val n = buf.int
        return ParsedHeader(version, dhPubRaw, pn, n)
    }

    data class ParsedMessage(
        val header: ByteArray,
        val nonce: ByteArray,       // 12 bytes
        val ciphertext: ByteArray,
        val tag: ByteArray          // 16 bytes
    )

    /** Build wire message: headerLen(4 BE) + header + nonce(12) + ciphertext + tag(16) */
    fun buildMessage(header: ByteArray, nonce: ByteArray, ciphertext: ByteArray, tag: ByteArray): ByteArray {
        val buf = ByteBuffer.allocate(4 + header.size + nonce.size + ciphertext.size + tag.size)
            .order(ByteOrder.BIG_ENDIAN)
        buf.putInt(header.size)
        buf.put(header)
        buf.put(nonce)
        buf.put(ciphertext)
        buf.put(tag)
        return buf.array()
    }

    fun parseMessage(data: ByteArray): ParsedMessage {
        require(data.size >= 4) { "Message too short" }
        val buf = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
        val headerLen = buf.int
        require(data.size >= 4 + headerLen + 12 + 16) { "Message too short for content" }

        val header = ByteArray(headerLen)
        buf.get(header)
        val nonce = ByteArray(12)
        buf.get(nonce)

        // Remaining = ciphertext + tag(16)
        val remaining = data.size - buf.position()
        val ctLen = remaining - 16
        val ciphertext = ByteArray(ctLen)
        buf.get(ciphertext)
        val tag = ByteArray(16)
        buf.get(tag)

        return ParsedMessage(header, nonce, ciphertext, tag)
    }
}
